File: ioi2020-mushrooms/solution/parallel-one-by-one-linear.java
/*
 * A parallel implementation of the one-by-one solution (creating threads linearly).
 * The purpose of the parallel solutions is to verify if the graders have the correct behavior
 *   when contestants use parallelism in their solutions.
 *
 * Author: Kian Mirjalali
 */
public class mushrooms {

	private int use_machine(int[] x) {
		return stub.use_machine(x);
	}

	public int count_mushrooms(int n) {
		final int[] species = new int[n];
		species[0] = 0;
		final int max_threads = 200;
		final int block_size = Math.max(1, n/max_threads);
		java.util.List<Thread> threads = new java.util.ArrayList<>();
		for (int begin_ = 1; begin_ < n; begin_ += block_size) {
			final int begin = begin_;
			final int end = Math.min(n, begin+block_size);
			threads.add(new Thread(() -> {
				for (int i = begin; i < end; i++)
					species[i] = use_machine(new int[] {0, i});
			}));
		}
		for (Thread th : threads)
			th.start();
		for (Thread th : threads)
			try {
				th.join();
			} catch(InterruptedException e) {
			}
		return (int)(java.util.Arrays.stream(species).filter(t -> (t == 0)).count());
	}
}

File: ioi2020-mushrooms/solution/parallel-one-by-one-linear.py
#
# A parallel implementation of the one-by-one solution (creating threads linearly).
# The purpose of the parallel solutions is to verify if the graders have the correct behavior
#   when contestants use parallelism in their solutions.
#
# Author: Kian Mirjalali
#
from typing import Final, List
from threading import Thread
from stub import use_machine

def count_mushrooms(n: int) -> int:
    species: Final[List[int]] = n * [0]
    max_threads: Final[int] = 200
    block_size: Final[int] = max(1, n//max_threads)
    threads: List[Thread] = []
    def action(begin: int, end: int):
        for i in range(begin, end):
            species[i] = use_machine([0, i])
    for begin in range(1, n, block_size):
        end: int = min(n, begin+block_size)
        threads.append(Thread(target=action, args=(begin, end)))
    for th in threads:
        th.start()
    for th in threads:
        th.join()
    return species.count(0)
